chore: move db defaulting logic out of the low level db interface

blaggacao opened the pull request:
This PR normalizes the configuration defaulting interface.

Configuration concerns (here: defaulting) should not bleed into the low level implementation of the database interface, even if it is database configuration
that is defaulted.

By moving any such database related configuration defaulting at least to the
calling site, this PR not only clarifies the business logic but also ~significantly~ improves the functional purity of the database interface (a bit).

blaggacao commented:
@ankush I rebased and fixed the lint (only failing test). Sorry for the extra round, I'm just getting to grips with the workflows over here. :smile:

blaggacao committed:
fix: procure db config from single authority

ensures that configuration is uniformely procured from local.conf
instead of making use of hard to audit multilevel fallback logic

Implementation Note:
- `get_db(host, port, user, password)` was stripped of any optional
  argument and therefrom all errors where fixed.
- All occurances of `grep 'frappe.db.db_'` where changed to
  `frappe.conf.db_`

blaggacao committed:
fix: procure db config from single authority

ensures that configuration is uniformely procured from local.conf
instead of making use of hard to audit multilevel fallback logic

Implementation Note:
- `get_db(host, port, user, password)` was stripped of any optional
  argument and therefrom all errors where fixed.
- All occurances of `grep 'frappe.db.db_'` where changed to
  `frappe.conf.db_`

blaggacao committed:
fix: procure db config from single authority

ensures that configuration is uniformely procured from local.conf
instead of making use of hard to audit multilevel fallback logic

Implementation Note:
- `get_db(host, port, user, password)` was stripped of any optional
  argument and therefrom all errors where fixed.
- All occurances of `grep 'frappe.db.db_'` where changed to
  `frappe.conf.db_`

blaggacao committed:
fix: procure db config from single authority

ensures that configuration is uniformely procured from local.conf
instead of making use of hard to audit multilevel fallback logic

Implementation Note:
- `get_db(host, port, user, password)` was stripped of any optional
  argument and therefrom all errors where fixed.
- All occurances of `grep 'frappe.db.db_'` where changed to
  `frappe.conf.db_`

blaggacao committed:
fix: procure db config from single authority

ensures that configuration is uniformely procured from local.conf
instead of making use of hard to audit multilevel fallback logic

Implementation Note:
- `get_db(host, port, user, password)` was stripped of any optional
  argument and therefrom all errors where fixed.
- All occurances of `grep 'frappe.db.db_'` where changed to
  `frappe.conf.db_`

blaggacao commented on frappe/database/database.py:
```diff
@@ -79,26 +79,19 @@ class InvalidColumnName(frappe.ValidationError):
-		self.user = user or frappe.conf.db_name
-		self.db_name = frappe.conf.db_name
+		self.host = host
+		self.port = port
+		self.user = user
+		self.cur_db_name = cur_db_name
```
The core of this PR: do not (quite implicitly) default on the database interface. This change requires being explicit at the calling site.

blaggacao committed:
fix: if no values provided, fall back to mysql/psql lib defaults

blaggacao commented on frappe/database/mariadb/database.py:
```diff
@@ -114,18 +114,23 @@ def set_execution_timeout(self, seconds: int):
 		self.sql("set session max_statement_time = %s", int(seconds))
 
 	def get_connection_settings(self) -> dict:
+		assert (
+			self.password
+		), "without unix domain socket connection, mariadb db password is currently reqired"
```
This assert already hints on how this prepares for better developer UX with passwordless development workflows in the local context via unix domain sockets.

blaggacao commented on frappe/database/db_manager.py:
```diff
@@ -70,16 +70,16 @@ def restore_database(target, source, user, password):
 			user=esc(user),
 			password=esc(password),
-			host=esc(frappe.db.host),
+			host=esc(frappe.conf.db_host or "127.0.0.1"),
```
This defaulting logic will be better handled by #21311

blaggacao commented on frappe/database/__init__.py:
```diff
@@ -39,14 +39,18 @@ def drop_user_and_database(db_name, root_login=None, root_password=None):
+		)
 	else:
 		import frappe.database.mariadb.database
 
-		return frappe.database.mariadb.database.MariaDBDatabase(host, user, password, port=port)
+		return frappe.database.mariadb.database.MariaDBDatabase(
+			host, user, password, port, cur_db_name=user
+		)
```
`cur_db_name=user` default now happens on a higher level interface.

blaggacao commented on frappe/database/mariadb/database.py:
```diff
@@ -114,18 +114,23 @@ def set_execution_timeout(self, seconds: int):
-			conn_settings["database"] = self.user
+			conn_settings["database"] = self.cur_db_name
 
+		if self.host:
+			conn_settings["host"] = self.host
+		if self.password:
+			conn_settings["password"] = self.password
```
host and password are not strictly mandatory for a mariadb connection (e.g. when using `MYSQL_UNIX_PORT` and when `whoami` is accordingly setup in the database.

This change anticipates this variant.

blaggacao replied:
Correction, `pymysql` doesn't actually do that but a more faithful lib implementation probably would.

blaggacao commented on frappe/tests/test_commands.py:
```diff
@@ -533,8 +533,8 @@ def test_backup_fails_with_exit_code(self):
 			frappe.conf.db_password + "INCORRECT PASSWORD",
-			db_host=frappe.db.host,
-			db_port=frappe.db.port,
+			db_host=frappe.conf.db_host,
+			db_port=frappe.conf.db_port,
```
these and similar &mdash; if `None` &mdash; are now taking the defaults from `libpg` and [`pymysql` here](https://github.com/PyMySQL/PyMySQL/blob/f4c348fdcf4ac21a92be58b6f94e9d7a13826a38/pymysql/connections.py#L291-L292). Delegating this to the upstream implementation makes frappe's defaulting implementation transparent to it.

blaggacao commented:
@ankush Tests are green and this is now ready for your / your team's consideration. Preferably before considering #21311, at all.

phot0n commented:
TODO:

- [ ] check for replica setup with different creds for authenticating with replica db server

blaggacao commented:
@phot0n Just to clarify, is that a TODO that I could help with? I don't have a replica setup at hand for testing, but maybe I can set one up?

agritheory commented:
@blaggacao Hola David, I can help you get one set up, probably easiest with a docker development instance. This isn't spec'd in any CI or test suite I'm aware of in the Frappe ecosystem, so I don't know that there's an canonical approach. Are you on the Spanish language ERPNext Telegram?

blaggacao commented:
@agritheory Something like:

1 Set up remote db
2 restore backup
3 configure `site_config.json`

?

agritheory commented:
@blaggacao That would be a good first test (remove vs local), the documentation for Amazon's RDS is a little dated but should help you get started. https://github.com/frappe/frappe/wiki/Using-Frappe-with-Amazon-RDS-(or-any-other-DBaaS). What I'm not clear about is if the ToDo is supposed to test against a Galera cluster, which is how I read it. https://mariadb.com/kb/en/what-is-mariadb-galera-cluster/

blaggacao commented:
## Test Protocol

### Test subject

https://github.com/blaggacao/frappe/blob/3cb01532637d023502c4f49fc0e4ada7040baab4/frappe/__init__.py#L280-L295
```py
def connect_replica():
	from frappe.database import get_db

	user = local.conf.db_name
	password = local.conf.db_password
	port = local.conf.replica_db_port

	if local.conf.different_credentials_for_replica:
		user = local.conf.replica_db_name
		password = local.conf.replica_db_password

	local.replica_db = get_db(host=local.conf.replica_host, user=user, password=password, port=port)

	# swap db connections
	local.primary_db = local.db
	local.db = local.replica_db
```

### Setup

```console
frappe on  refactor-default-heuristic-current-db [$?] via  v18.14.0 via 🐍 v3.10.9 (env) took 6m21s
❯ pwd
/home/blaggacao/src/gitlab.com/pristina/infraestructura/bench/apps/frappe

❯ export TYPE="server"
❯ export GITHUB_WORKSPACE=`pwd`
❯ export DB="mariadb"
❯ start-mariadb  # private bench-nix script for mariadb w/ correct settings and signal handler
❯ bash -x .github/helper/install.sh
❯ docker run -e MARIADB_ROOT_PASSWORD=travis -P mariadb
> # ^^ the replica: tcp socket ends up being 0.0.0.0:49155 (but changes on occasional restarts further down)
❯ mariadb --host 127.0.0.1 --port 49155 -u root -ptravis -e "SET collation_server = 'utf8mb4_unicode_ci'";
❯ mariadb --host 127.0.0.1 --port 49155 -u root -ptravis -e "SET character_set_server = 'utf8mb4'";
❯ cd ~/frappe-bench/sites/test_site
❯ . env/bin/activate
❯ pip install frappe-bench
❯ nix shell nixpkgs#mariadb # for mysql and mysqldump bins
❯ bench --site test_site backup
Backup Summary for test_site at 2023-06-17 20:47:27.448984
Config  : ./test_site/private/backups/20230617_204727-test_site-site_config_backup.json 591.0B
Database: ./test_site/private/backups/20230617_204727-test_site-database.sql.gz         251.7KiB
Backup for Site test_site has been successfully completed
# pointing site_config to new port
❯ bench --site test_site restore --mariadb-root-password="travis" ./test_site/private/backups/20230617_204727-test_site-database.sql.gz
# pointing site_config back to old port
❯ nix shell nixpkgs#redis # for redis-server command
```
---

### Case 1: "same credentials as in primary db / maintenance mode"

```diff
diff --git a/site_config.json.bkp b/site_config.json
index e86e701..70e254d 100644
--- a/site_config.json.bkp
+++ b/site_config.json
@@ -3,6 +3,11 @@
     "db_port": 3306,
     "db_name": "test_frappe",
     "db_password": "test_frappe",
+    "replica_host": "127.0.0.1",
+    "replica_db_port": 49155,
+    "read_from_replica": true,
+    "maintenance_mode": true,
+    "allow_reads_during_maintenance": true,
     "allow_tests": true,
     "db_type": "mariadb",
     "auto_email_id": "test@example.com",
```
**Results**


Result with replica mysql docker stopped:
```
pymysql.err.OperationalError: (2003, "Can't connect to MySQL server on '127.0.0.1' ([Errno 111] Connection refused)")
```
Result with replica mysql docker running:
![image](https://github.com/frappe/frappe/assets/7548295/7d22cf6c-1695-4231-a091-01340b608d88)

---
### Case 2: "_different_ credentials as in primary db / maintenance mode"

```diff
diff --git a/site_config.json.bkp b/site_config.json
index e86e701..502bff6 100644
--- a/site_config.json.bkp
+++ b/site_config.json
@@ -1,8 +1,16 @@
 {
     "db_host": "127.0.0.1",
     "db_port": 3306,
-    "db_name": "test_frappe",
-    "db_password": "test_frappe",
+    "db_name": "invalid",
+    "db_password": "invalid",
+    "replica_host": "127.0.0.1",
+    "replica_db_port": 49156,
+    "different_credentials_for_replica": true,
+    "replica_db_name": "test_frappe",
+    "replica_db_password": "test_frappe",
+    "read_from_replica": true,
+    "maintenance_mode": true,
+    "allow_reads_during_maintenance": true,
     "allow_tests": true,
     "db_type": "mariadb",
     "auto_email_id": "test@example.com",
```

Result (here the public website):
![image](https://github.com/frappe/frappe/assets/7548295/86bebf8a-71e4-49e6-9961-0da7469f8de0)

---

### Case 3: "client side read-replica based load balancing"

```diff
diff --git a/site_config.json.bkp b/site_config.json
index e86e701..a72125d 100644
--- a/site_config.json.bkp
+++ b/site_config.json
@@ -3,6 +3,9 @@
     "db_port": 3306,
     "db_name": "test_frappe",
     "db_password": "test_frappe",
+    "replica_host": "127.0.0.1",
+    "replica_db_port": 49156,
+    "read_from_replica": true,
     "allow_tests": true,
     "db_type": "mariadb",
     "auto_email_id": "test@example.com",
```

Finding read-balanced endpoints:
```console
❯ rg 'read_only\(' -tpy
frappe/__init__.py
791:def read_only():

frappe/desk/query_report.py
72:@frappe.read_only()
179:@frappe.read_only()

frappe/desk/notifications.py
22:@frappe.read_only()
244:@frappe.read_only()

frappe/desk/reportview.py
20:@frappe.read_only()
33:@frappe.read_only()
48:@frappe.read_only()
334:@frappe.read_only()
500:@frappe.read_only()
516:@frappe.read_only()

frappe/desk/desktop.py
387:@frappe.read_only()
```

Choosing (i.e. load a workspace)
```py
@frappe.whitelist()
@frappe.read_only()
def get_desktop_page(page):
   ...
```

Result: `get_desktop_page` loading normally with 200:
(after having manually completed onboarding)
![image](https://github.com/frappe/frappe/assets/7548295/9ae237f2-fbdc-4034-8dcc-5c9cc4ea4732)

---
@phot0n These tests cover all possible code paths:
- different replica credentials
- client side read balancing
- read-only from replica during maintenance mode

blaggacao committed:
fix: replace field replacing correct dom element

blaggacao committed:
fix: replace_field

splice needs an (integer) index not an element

also section field management must be taken into account for all sections

blaggacao commented:
@phot0n I rebased this PR to ensure a clean history and to re-run the tests.

shariquerik committed:
fix: avoid this line if there is no tab break field in the form

shariquerik committed:
Merge branch 'develop' into fix-field-replace

blaggacao committed:
fix: gently log the port on which the realtime service listens

without this, the service remains completely silent (e.g. in overmind) and it is hard to tell if it is even running

blaggacao committed:
fix: checkpoint the supported schemes for connectivity

This PR implements a gateway + error that clearly hints the operator at
a misconfigured system during runtime.

Particularity, against the multiple library-provided ways of configuring
redis connection strings (in python), this hard stops if an unsupported
one is chosen by accident.

blaggacao committed:
fix: add env overrides for service orchestration

This avoids having to manipulate config files in brittle bash
entrypoints that need to react to dynamic service discovery.

This significantly improves the operability of various bench sites.

blaggacao committed:
fix: add env overrides for service orchestration

This avoids having to manipulate config files in brittle bash
entrypoints that need to react to dynamic service discovery.

This significantly improves the operability of various bench sites.

blaggacao committed:
fix: add env overrides for service orchestration

This avoids having to manipulate config files in brittle bash
entrypoints that need to react to dynamic service discovery.

This significantly improves the operability of various bench sites.

blaggacao committed:
Merge branch 'fix-add-checkpoint-on-connectivity-interface' into 00-likely-accepted-fixes

blaggacao committed:
Merge branch 'fix-field-replace' into 00-likely-accepted-fixes

blaggacao committed:
Merge branch 'fix-log-port-on-which-realtime-service-listens' into 00-likely-accepted-fixes

blaggacao committed:
Merge branch 'fix-options-from-env-variables' into 00-likely-accepted-fixes

blaggacao committed:
Merge branch 'fix-procure-db-config-from-single-authority' into 00-likely-accepted-fixes

blaggacao committed:
fix: set defaults at the highest level interface

Having defaults burried far away from the actual executable
entrypoint is significantly harder to operate.

blaggacao committed:
Merge branch 'fix-procure-db-config-from-single-authority' into 00-likely-accepted-fixes

blaggacao commented:
Putting this on-hold. This becomes actually a trivial PR with the expanded guarantees of https://github.com/frappe/frappe/pull/21577 and https://github.com/frappe/frappe/pull/21578

blaggacao committed:
Merge branch 'fix-procure-db-config-from-single-authority' into 00-likely-accepted-fixes

blaggacao committed:
Merge branch 'fix-options-from-env-variables' into 00-likely-accepted-fixes

blaggacao committed:
Merge branch 'fix-procure-db-config-from-single-authority' into 00-likely-accepted-fixes

blaggacao committed:
Merge branch 'fix-procure-db-config-from-single-authority' into 00-likely-accepted-fixes
